docs/tutorial/Tutorial.md

```diff
@@ -0,0 +1,501 @@
+           offset: isize,
+           semantics: AllocationSemantics,
+       ) -> Address {
+       };
+       fn post_copy(
+           &mut self,
+           _obj: ObjectReference,
+           _tib: Address,
+           _bytes: usize,
+           _semantics: AllocationSemantics,
+       ) {
+       }
+       ```
+   3. To `new`, add an initialiser for the class:
+       ```rust
+       Self {
+             plan: &mmtk.plan,
+             mygc: BumpAllocator::new(OpaquePointer::UNINITIALIZED, None, &mmtk.plan),
+         }
+       ```
+   4. In `init`, set the `tls` variable in the held instance of `mygc` to the one passed to the function.
+   5. In `prepare`, rebind the allocator to the tospace.
+   6. Leave `release` with an empty body.
+   7. In `alloc`, call the allocator's `alloc` function. Above the function, use an inline attribute (`#[inline(always)]`) to tell the Rust compiler to always inline the function.
+   8. In `post_copy` add the following code. Also, add an inline (always) attribute.
+       ```rust
+       forwarding_word::clear_forwarding_bits::<VM>(obj);
+       ```
+5. Add a new public structure, `MyGCProcessEdges`, with the type parameter `<VM:VMBinding>`. It will hold an instance of `ProcessEdgesBase` and `PhantomData`, and implement the Default trait:
+    ```rust
+    #[derive(Default)]
+    pub struct MyGCProcessEdges<VM: VMBinding> {
+        base: ProcessEdgesBase<MyGCProcessEdges<VM>>,
+        phantom: PhantomData<VM>,
+    }
+    ```
+6. Add a new implementations block `impl<VM:VMBinding> ProcessEdgesWork for MyGCProcessEdges<VM>`.
+   1. Add a VM type alias (`type VM = VM`).
+   2. Add a new method, `new`.
+       ```rust
+       fn new(edges: Vec<Address>, _roots: bool) -> Self {
+           Self {
+               base: ProcessEdgesBase::new(edges),
+               ..Default::default()
+           }
+       }
+      ```
+   3. Add a new method, `trace_object(&mut self, object: ObjectReference)`.
+     1. This method should return an ObjectReference, and use the inline (*not* always) attribute.
+     2. Check if the object passed into the function is null (`object.is_null()`). If it is, return the object.
+     3. Check if the object is in the tospace (`self.plan().tospace().in_space(object)`). If it is, call `trace_object` through the tospace to check if the object is alive, and return the result:
+         ```rust
+         self.plan().tospace().trace_object(
+               self,
+               object,
+               super::global::ALLOC_MyGC,
+               self.worker().local(),
+           )
+         ```
+     4. If it is not in the tospace, check if the object is in the fromspace and return the result of the fromspace's `trace_object` if it is.
+     5. If it is in neither space, it must be in the immortal space, or large object space. Trace the object with `self.plan().common.trace_object(self, object)`.
+7. Add two new implementation blocks, `Deref` and `DerefMut` for `MyGCProcessEdges`. 
```

Comment:
This allows dereferencing `MyGCProcessEdges` to `ProcessEdgesBase`, and allows easy access to fields in `ProcessEdgesBase`.